Set Storage Policy ID while registering migrated in-tree vSphere volume

divyenpatel opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is supplying SPBM Policy of in-tree vSphere Volume while registering volume with CNS.


**Which issue this PR fixes**
fixes https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/255

**Special notes for your reviewer**:

Verified that PV Created using Storage Policy `vSAN Default Storage Policy` and migrated to vSphere CSI Driver and showing up on the CNS UI with SPBM Policy associated.

```
# kubectl describe pv 
Name:            pvc-b1792c12-35f1-4cfd-946c-7ef84643cdd4
Labels:          <none>
Annotations:     kubernetes.io/createdby: vsphere-volume-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller: yes
                 pv.kubernetes.io/migrated-to: csi.vsphere.vmware.com
                 pv.kubernetes.io/provisioned-by: kubernetes.io/vsphere-volume
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    vcpsc
Status:          Bound
Claim:           default/vcppvc
Reclaim Policy:  Delete
Access Modes:    RWO
VolumeMode:      Filesystem
Capacity:        2Gi
Node Affinity:   <none>
Message:         
Source:
    Type:               vSphereVolume (a Persistent Disk resource in vSphere)
    VolumePath:         [vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-b1792c12-35f1-4cfd-946c-7ef84643cdd4.vmdk
    FSType:             ext4
    StoragePolicyName:  vSAN Default Storage Policy
Events:                 <none>
```

<img width="1298" alt="Screen Shot 2020-07-27 at 12 23 47 PM" src="https://user-images.githubusercontent.com/22985595/88583616-52282d00-d005-11ea-8eae-20a9d4aa29f5.png">



**Testing done with on latest commit**

1. Created PV using storage class having StoragePolicyName set to `my-custom-vsan-policy`

```
# kubectl describe pv
Name:            pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7
Labels:          <none>
Annotations:     kubernetes.io/createdby: vsphere-volume-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller: yes
                 pv.kubernetes.io/provisioned-by: kubernetes.io/vsphere-volume
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    vcpsc
Status:          Bound
Claim:           default/vcppvc
Reclaim Policy:  Delete
Access Modes:    RWO
VolumeMode:      Filesystem
Capacity:        2Gi
Node Affinity:   <none>
Message:         
Source:
    Type:               vSphereVolume (a Persistent Disk resource in vSphere)
    VolumePath:         [vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk
    FSType:             ext4
    StoragePolicyName:  my-custom-vsan-policy
Events:                 <none>
```

2. Renamed SPBM Policy to my-custom-vsan-policy-1 on vCenter. Enabled Migration and Installed CSI Driver. Verified volume registration is failing with Profile not found.

> 2020-07-30T18:06:32.392762259Z 2020-07-30T18:06:32.392Z	ERROR	migration/migration.go:369	Error occurred while getting stroage policy ID from storage policy name: "my-custom-vsan-policy", err: no pbm profile found with name: "my-custom-vsan-policy"	{"TraceId": "31b1c874-d132-4093-b5c3-c8abde6ec8c3"}

> 2020-07-30T18:06:32.393344986Z 2020-07-30T18:06:32.392Z	ERROR	migration/migration.go:186	failed to register volume for volumeSpec: {[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk my-custom-vsan-policy}, with err: Error occurred while getting stroage policy ID from storage policy name: "my-custom-vsan-policy", err: no pbm profile found with name: "my-custom-vsan-policy"	{"TraceId": "31b1c874-d132-4093-b5c3-c8abde6ec8c3"}



3. Renamed SPBM Policy back to to `my-custom-vsan-policy` on vCenter.

4. Performed Operation on volume and Verified volume registration happened successfully and metadata is updated in CNS.

```
kubectl label pv pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7 testkey=testval
persistentvolume/pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7 labeled
```


Logs

```
2020-07-30T18:08:35.318511445Z 2020-07-30T18:08:35.318Z	INFO	migration/migration.go:183	Could not retrieve VolumeID from cache for Volume Path: "[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk". volume may not be registered. Registering Volume with CNS	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:35.366211058Z 2020-07-30T18:08:35.365Z	DEBUG	migration/migration.go:365	Obtaining storage policy ID for storage policy name: "my-custom-vsan-policy"	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:35.429188686Z 2020-07-30T18:08:35.428Z	INFO	migration/migration.go:372	Obtained storage policy ID: "7ee29319-56cb-4b56-bef3-36bfd01d2f62" for storage policy name: "my-custom-vsan-policy"	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:35.429224951Z 2020-07-30T18:08:35.428Z	INFO	migration/migration.go:380	Registering volume: "[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk" using backingDiskURLPath :"https://10.186.32.129/folder/56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk?dcPath=VSAN-DC&dsName=vsanDatastore"	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:35.436661290Z 2020-07-30T18:08:35.436Z	DEBUG	migration/migration.go:401	vSphere CNS driver registering volume "[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk" with create spec (*types.CnsVolumeCreateSpec)(0xc000257180)({
2020-07-30T18:08:35.436700425Z  DynamicData: (types.DynamicData) {
2020-07-30T18:08:35.436706054Z  },
2020-07-30T18:08:35.436710315Z  Name: (string) (len=36) "af67ba73-d28f-11ea-afea-76f74793fadd",
2020-07-30T18:08:35.436715185Z  VolumeType: (string) (len=5) "BLOCK",
2020-07-30T18:08:35.436718202Z  Datastores: ([]types.ManagedObjectReference) <nil>,
2020-07-30T18:08:35.436721184Z  Metadata: (types.CnsVolumeMetadata) {
2020-07-30T18:08:35.436723845Z   DynamicData: (types.DynamicData) {
2020-07-30T18:08:35.436726440Z   },
2020-07-30T18:08:35.436728981Z   ContainerCluster: (types.CnsContainerCluster) {
2020-07-30T18:08:35.436731766Z    DynamicData: (types.DynamicData) {
2020-07-30T18:08:35.436734386Z    },
2020-07-30T18:08:35.436737046Z    ClusterType: (string) (len=10) "KUBERNETES",
2020-07-30T18:08:35.436739817Z    ClusterId: (string) (len=15) "demo-cluster-id",
2020-07-30T18:08:35.436742655Z    VSphereUser: (string) (len=27) "Administrator@vsphere.local",
2020-07-30T18:08:35.436745453Z    ClusterFlavor: (string) (len=7) "VANILLA"
2020-07-30T18:08:35.436748061Z   },
2020-07-30T18:08:35.436750462Z   EntityMetadata: ([]types.BaseCnsEntityMetadata) <nil>,
2020-07-30T18:08:35.436753377Z   ContainerClusterArray: ([]types.CnsContainerCluster) (len=1 cap=1) {
2020-07-30T18:08:35.436756042Z    (types.CnsContainerCluster) {
2020-07-30T18:08:35.436758790Z     DynamicData: (types.DynamicData) {
2020-07-30T18:08:35.436761517Z     },
2020-07-30T18:08:35.436764133Z     ClusterType: (string) (len=10) "KUBERNETES",
2020-07-30T18:08:35.436782608Z     ClusterId: (string) (len=15) "demo-cluster-id",
2020-07-30T18:08:35.436785487Z     VSphereUser: (string) (len=27) "Administrator@vsphere.local",
2020-07-30T18:08:35.436788106Z     ClusterFlavor: (string) (len=7) "VANILLA"
2020-07-30T18:08:35.436790746Z    }
2020-07-30T18:08:35.436793143Z   }
2020-07-30T18:08:35.436795972Z  },
2020-07-30T18:08:35.436798559Z  BackingObjectDetails: (*types.CnsBlockBackingDetails)(0xc0005931a0)({
2020-07-30T18:08:35.436801320Z   CnsBackingObjectDetails: (types.CnsBackingObjectDetails) {
2020-07-30T18:08:35.436804092Z    DynamicData: (types.DynamicData) {
2020-07-30T18:08:35.436806667Z    },
2020-07-30T18:08:35.436809136Z    CapacityInMb: (int64) 0
2020-07-30T18:08:35.436811637Z   },
2020-07-30T18:08:35.436813963Z   BackingDiskId: (string) "",
2020-07-30T18:08:35.436816791Z   BackingDiskUrlPath: (string) (len=166) "https://10.186.32.129/folder/56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk?dcPath=VSAN-DC&dsName=vsanDatastore"
2020-07-30T18:08:35.436820177Z  }),
2020-07-30T18:08:35.436822745Z  Profile: ([]types.BaseVirtualMachineProfileSpec) (len=1 cap=1) {
2020-07-30T18:08:35.436825531Z   (*types.VirtualMachineDefinedProfileSpec)(0xc000804fc0)({
2020-07-30T18:08:35.436828127Z    VirtualMachineProfileSpec: (types.VirtualMachineProfileSpec) {
2020-07-30T18:08:35.436830701Z     DynamicData: (types.DynamicData) {
2020-07-30T18:08:35.436833189Z     }
2020-07-30T18:08:35.436835608Z    },
2020-07-30T18:08:35.436838977Z    ProfileId: (string) (len=36) "7ee29319-56cb-4b56-bef3-36bfd01d2f62",
2020-07-30T18:08:35.436841968Z    ReplicationSpec: (*types.ReplicationSpec)(<nil>),
2020-07-30T18:08:35.436844780Z    ProfileData: (*types.VirtualMachineProfileRawData)(<nil>),
2020-07-30T18:08:35.436847658Z    ProfileParams: ([]types.KeyValue) <nil>
2020-07-30T18:08:35.436850296Z   })
2020-07-30T18:08:35.436852686Z  },
2020-07-30T18:08:35.436855059Z  CreateSpec: (types.BaseCnsBaseCreateSpec) <nil>
2020-07-30T18:08:35.436857702Z })
2020-07-30T18:08:35.436860359Z 	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:35.448081801Z 2020-07-30T18:08:35.447Z	DEBUG	volume/manager.go:167	Update VSphereUser from Administrator@vsphere.local to VSPHERE.LOCAL\Administrator	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.755911435Z 2020-07-30T18:08:40.745Z	INFO	volume/manager.go:201	CreateVolume: VolumeName: "af67ba73-d28f-11ea-afea-76f74793fadd", opId: "b4860e92"	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.756430556Z 2020-07-30T18:08:40.754Z	DEBUG	volume/manager.go:244	Deleted task for af67ba73-d28f-11ea-afea-76f74793fadd from volumeTaskMap for statically provisioned volume	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.756557864Z 2020-07-30T18:08:40.754Z	INFO	volume/manager.go:249	CreateVolume: Volume created successfully. VolumeName: "af67ba73-d28f-11ea-afea-76f74793fadd", opId: "b4860e92", volumeID: "6dbf2996-e344-436b-92a1-7ace86bf56f6"	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.756565044Z 2020-07-30T18:08:40.755Z	INFO	migration/migration.go:410	successfully registered volume "[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk" as container volume with ID: "6dbf2996-e344-436b-92a1-7ace86bf56f6"	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.756574658Z 2020-07-30T18:08:40.755Z	INFO	migration/migration.go:189	successfully registered volumeSpec: {[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk my-custom-vsan-policy} with CNS. VolumeID: 6dbf2996-e344-436b-92a1-7ace86bf56f6	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.757387713Z 2020-07-30T18:08:40.757Z	DEBUG	migration/migration.go:197	Saving cnsvSphereVolumeMigration CR: {{ } {6dbf2996-e344-436b-92a1-7ace86bf56f6      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk 6dbf2996-e344-436b-92a1-7ace86bf56f6}}	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
2020-07-30T18:08:40.758106192Z 2020-07-30T18:08:40.757Z	INFO	migration/migration.go:268	creating CR for cnsVSphereVolumeMigration: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:6dbf2996-e344-436b-92a1-7ace86bf56f6 GenerateName: Namespace: SelfLink: UID: ResourceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName: ManagedFields:[]} Spec:{VolumePath:[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk VolumeID:6dbf2996-e344-436b-92a1-7ace86bf56f6}}	{"TraceId": "f9ea8b17-fa7d-4207-b8b4-a51cae8e29e5"}
```


```
# kubectl get Cnsvspherevolumemigrations
NAME                                   AGE
6dbf2996-e344-436b-92a1-7ace86bf56f6   49s

# kubectl describe  Cnsvspherevolumemigrations
Name:         6dbf2996-e344-436b-92a1-7ace86bf56f6
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  cns.vmware.com/v1alpha1
Kind:         CnsVSphereVolumeMigration
Metadata:
  Creation Timestamp:  2020-07-30T18:08:40Z
  Generation:          1
  Managed Fields:
    API Version:  cns.vmware.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:volumeid:
        f:volumepath:
    Manager:         vsphere-syncer
    Operation:       Update
    Time:            2020-07-30T18:08:40Z
  Resource Version:  1495126
  Self Link:         /apis/cns.vmware.com/v1alpha1/cnsvspherevolumemigrations/6dbf2996-e344-436b-92a1-7ace86bf56f6
  UID:               28efc510-7a49-4293-9b3d-43e0c266658b
Spec:
  Volumeid:    6dbf2996-e344-436b-92a1-7ace86bf56f6
  Volumepath:  [vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-a0481a96-5d09-49f1-98cd-9f30bbf6d7d7.vmdk
Events:        <none>
```

**E2E Test Execution Logs and Result**

exected all block e2e tests.
Test log - https://gist.github.com/divyenpatel/bc07b5d2e31a9185c9a444e1302a528b

1 test failed out of 33 and failure is not related to this change.

```
Summarizing 1 Failure:

[Fail] [csi-block-vanilla] Relocate detached volume  [BeforeEach] Verify relocating detached volume works fine 
/Users/divyenp/go/src/github.com/divyenpatel/vsphere-csi-driver/tests/e2e/e2e_common.go:121

Ran 33 of 106 Specs in 6217.758 seconds
FAIL! -- 32 Passed | 1 Failed | 0 Pending | 73 Skipped
--- FAIL: TestE2E (6217.77s)
FAIL
```




**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Set Storage Policy ID while registering migrated in-tree vSphere volume
```

@chethanv28 Can you help review this PR?

svcbot-qecnsdp commented:
Can one of the admins verify this patch?

chethanv28 commented on pkg/common/unittestcommon/types.go:
```diff
@@ -43,7 +44,7 @@ type mockVolumeMigration struct {
 	// GetVolumeID returns VolumeID for given VolumePath
 	// Returns an error if not able to retrieve VolumeID.
-	GetVolumeID(ctx context.Context, volumePath string) (string, error)
+	GetVolumeID(ctx context.Context, volumeSpec migration.VolumeSpec) (string, error)
```
Nit: Update doc for method. GetVolumeID returns VolumeID for a given migration volumeSpec

divyenpatel replied:
fixed.

chethanv28 commented on pkg/apis/migration/migration.go:
```diff
@@ -47,12 +47,19 @@ import (
 
+// VolumeSpec contains VolumePath and StoragePolicyID
+// using which Volume can be looked up using VolumeMigrationService
+type VolumeSpec struct {
```
How about renaming VolumeSpec ==> MigrationVolumeSpec or VolumeMigrationSpec ?
Would it be too specific ?

divyenpatel replied:
`VolumeSpec` type is part of `migration` package.  So I think it is ok to not prefix or postfix with keyword `migration`.
what is your opinion? Is it creating confusion, with volume create spec, query volume spec ...

chethanv28 replied:
That's right! Caller will always migration.VolumeSpec. This makes sense.
We can keep it as VolumeSpec.

chethanv28 commented on pkg/syncer/fullsync.go:
```diff
@@ -53,7 +54,7 @@ func csiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) {
-			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, pv.Spec.VsphereVolume.VolumePath)
+			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, migration.VolumeSpec{VolumePath: pv.Spec.VsphereVolume.VolumePath, StoragePolicyID: pv.Spec.VsphereVolume.StoragePolicyID})
 			if err != nil {
 				log.Errorf("FullSync: Failed to get VolumeID from volumeMigrationService for volumePath: %s with error %+v", pv.Spec.VsphereVolume.VolumePath, err)
```
Don't we have to log StoragePolicyID in the error message ? Will that be useful for debugging ?

divyenpatel replied:
Yes, let me add the spec to help enhance debugging.

chethanv28 commented on pkg/syncer/fullsync.go:
```diff
@@ -134,7 +135,7 @@ func fullSyncCreateVolumes(ctx context.Context, createSpecArray []cnstypes.CnsVo
 		if metadataSyncer.coCommonInterface.IsFSSEnabled(ctx, common.CSIMigration) && pv.Spec.VsphereVolume != nil {
-			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, pv.Spec.VsphereVolume.VolumePath)
+			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, migration.VolumeSpec{VolumePath: pv.Spec.VsphereVolume.VolumePath, StoragePolicyID: pv.Spec.VsphereVolume.StoragePolicyID})
 			if err != nil {
```
Same as above

divyenpatel replied:
Yes I will add

chethanv28 commented:
Minor comments! Overall Looks good to me-ish

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -164,26 +171,26 @@ func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Man
 		return "", err
 	}
-	log.Infof("successfully registered volume: %q with CNS. VolumeID: %q", volumePath, volumeID)
+	log.Infof("successfully registered volume: %q with CNS. volumeSpec: %v", volumeSpec, volumeID)
```
Shouldn't this be: "successfully registered volumeSpec: %v with CNS. VolumeID: %q"

divyenpatel replied:
yes.

changed to

```
log.Infof("successfully registered volumeSpec: %v with CNS. VolumeID: %v", volumeSpec, volumeID)
```

SandeepPissay commented on pkg/csi/service/vanilla/controller.go:
```diff
@@ -693,7 +710,7 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 		// Migration feature switch is enabled
 		if volumePath != "" {
-			req.VolumeId, err = volumeMigrationService.GetVolumeID(ctx, volumePath)
+			req.VolumeId, err = volumeMigrationService.GetVolumeID(ctx, migration.VolumeSpec{VolumePath: volumePath})
```
Any reason why we dont set the StoragePolicyID here?

divyenpatel replied:
In `ControllerUnpublishVolume`, we do not get SPBM policy name as we get it in `ControllerPublishVolume` using `VolumeContext`.

**`ControllerUnpublishVolume` Request**

> 2020-07-28T18:37:49.932064332Z {"level":"info","time":"2020-07-28T18:37:49.92822473Z","caller":"vanilla/controller.go:691","msg":"ControllerUnpublishVolume: called with args {VolumeId:[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-b1792c12-35f1-4cfd-946c-7ef84643cdd4.vmdk NodeId:k8s-node2 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"1d56da6d-c0fd-4971-b8de-f2c33f5ffdc5"}


**`ControllerPublishVolume` Request**

> 2020-07-28T18:36:32.832106394Z {"level":"info","time":"2020-07-28T18:36:32.825082689Z","caller":"vanilla/controller.go:577","msg":"ControllerPublishVolume: called with args {VolumeId:[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-b1792c12-35f1-4cfd-946c-7ef84643cdd4.vmdk NodeId:k8s-node2 VolumeCapability:mount:<fs_type:\"ext4\" > access_mode:<mode:SINGLE_NODE_WRITER >  Readonly:false Secrets:map[] VolumeContext:map[storagepolicyname:vSAN Default Storage Policy] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"0f0b7fba-a6d2-49bb-86b6-d88afe67444a"}



Also by this time volume is already registered with CNS, and Pod is up and running. In `ControllerUnpublish` mostly we expect volume migration CR present in the ETCD and the in-memory cache is pre-populated with path to ID mapping.

SandeepPissay replied:
> Also by this time volume is already registered with CNS

This may not be true if the first call back we get for the volume after CSI migration is ControllerUnpublishVolume.

divyenpatel replied:
The very first call we get is in the syncer for PV update when migration annotation is added.

Also we have a node drain requirement for migration - https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md#node-drain-requirement to keep volume lifecycle workflow, either with in-tree or with CSI driver. 

We are not supporting the case, where volume is attached using in-tree VCP, but need to be detached with CSI Driver.

SandeepPissay replied:
Can we add some doc in ControllerUnpublishVolume to state that this will never be the first call back for vmdk registration with CNS for so and so reasons? This will help to understand why storage policy is optional in this case.

divyenpatel replied:
providing the following doc.

```
// ControllerUnpublishVolume will never be the first call back for vmdk registration with CNS.
// Here in the migration.VolumeSpec, we do not supply SPBM Policy name.
// Node drain is the pre-requisite for volume migration, so volume will be registered with SPBM policy
// during ControllerPublish if metadata-syncer fails to register volume using associated SPBM Policy.
// for ControllerUnpublishVolume we anticipate volume is already registered with CNS, and volumeMigrationService
// should return volumeID for requested VolumePath
```

SandeepPissay commented on pkg/syncer/fullsync.go:
```diff
@@ -53,7 +54,7 @@ func csiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) {
 		if metadataSyncer.coCommonInterface.IsFSSEnabled(ctx, common.CSIMigration) && pv.Spec.VsphereVolume != nil {
 			// For vSphere volumes, the migration service will register volumes in CNS.
-			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, pv.Spec.VsphereVolume.VolumePath)
+			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, migration.VolumeSpec{VolumePath: pv.Spec.VsphereVolume.VolumePath, StoragePolicyID: pv.Spec.VsphereVolume.StoragePolicyID})
```
What if the StoragePolicyID does not exist in SPBM? Would it make sense to ignore the storage policy ID for the CNS register operation?

I have the same questions for similar changes in this PR.

divyenpatel replied:
We get this StoragePolicyID from the PV spec.

```
"vsphereVolume": {
    "fsType": "ext4",
    "storagePolicyID": "aa6d5a82-1c88-45da-85d3-3d74b91a5bad",
     "storagePolicyName": "vSAN Default Storage Policy",
     "volumePath": "[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-b1792c12-35f1-4cfd-946c-7ef84643cdd4.vmdk"
}
```

and we expect this policy ID to remain present in the vSphere, as Volumes are already using this policy. 
This PV spec is also immutable, so we can not edit PV and specify a new policy or remove policy association.

We should not give the flexibility to the user to delete the SPBM policy when it is in use by PVs. 

If we want to make this optional then, when the policy is not present in the vCenter, sever throws a genetic CNS fault and we will need to catch this fault, parse the localized message and make a decision to recall create volume without supplying policy id, which is not optimal.

SandeepPissay replied:
> and we expect this policy ID to remain present in the vSphere, as Volumes are already using this policy.

SPBM remembers the policy association of disks that are attached. Once the disk is detached, SPBM does not know about the policy and will allow policy deletion. If the policy is deleted in SPBM, then this code change affects the metadata sync and full sync workflows.

divyenpatel replied:
So this is like the requirement of not rename datastore or alter VC inventory, as VCP is very sensitive to such changes.
SPBM policy also should not be deleted by storage admin, and if it is deleted, the volume can not be migrated.

if we have dedicated fault from CNS API to tell that specified policy is not present in VC, we can consider this requirement. Comparing localized messages in the fault will not be good.

SandeepPissay replied:
I'm not suggesting to compare error messages. We can invoke SPBM API to know if the policy exists. Any reason why not to do this check? If the policy exists then the policy will be applied for volume registration, if not then don't apply the policy. Wouldn't this fix the problem I raised?

divyenpatel replied:
I can add that extra VC call using PBM client that will satisfy this requirement. 

but one more thing to confirm - If VC admin has accidentally deleted SPBM policy but later with VC backup/restore if policy appears back in the VC, we can not apply back restored policy on the volume, without de-registering and re-registering back.

Re-registration of volume with CNS does not consider the policy parameters if the volume is already registered with CNS.

SandeepPissay replied:
As we discussed, lets use the "storagePolicyName" to load the policy Id and use this to register the vmdk as a CNS volume. If the policy is deleted in vCenter, the syncer will fail until the vSphere admin recreates the policy with the same name.

divyenpatel replied:
done. Addressed as part of - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/287/commits/5a0251a40550d0cc1141b86f040084a2353c80f4

divyenpatel commented:
@SandeepPissay @chethanv28 I have addressed your review comments in  https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/287/commits/af536faa51bc6bda79836c15ae129741c6a71a59

divyenpatel commented:
@SandeepPissay @chethanv28 I have squashed all commits and updated "Testing Done" on the PR.

Can you approve this PR?

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -353,13 +360,24 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+			log.Error(msg)
+			return "", errors.New(msg)
+		}
+		log.Infof("Obtained storage policy ID: %q for storage policy name: %q", storagePolicyID, volumeSpec.StoragePolicyName)
```
Info -> Debug

divyenpatel replied:
done.

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -374,18 +392,24 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		log.Infof("successfully registered volume %q as container volume with ID: %q", volumeSpec.VolumePath, volumeID.Id)
 	} else {
-		msg := fmt.Sprintf("registration failed for volumePath: %q", volumePath)
+		msg := fmt.Sprintf("registration failed for volumePath: %q", volumeSpec.VolumePath)
```
Can we add more context like Storage Policy name/Id?

divyenpatel replied:
In the message above we print all details with

```
log.Warnf("failed to register volume %q with createSpec: %v. error: %+v", volumeSpec.VolumePath, createSpec, err)
```

with the contextual tracing, we can get all details.

changing this to

```
msg := fmt.Sprintf("registration failed for volumeSpec: %v", volumeSpec)
```

SandeepPissay commented on pkg/common/cns-lib/vsphere/pbm.go:
```diff
@@ -82,6 +82,11 @@ func (vc *VirtualCenter) DisconnectPbm(ctx context.Context) error {
 // GetStoragePolicyIDByName gets storage policy ID by name.
 func (vc *VirtualCenter) GetStoragePolicyIDByName(ctx context.Context, storagePolicyName string) (string, error) {
 	log := logger.GetLogger(ctx)
+	err := vc.ConnectPbm(ctx)
```
Why can't this be part of VirtualCenter methods? This way the callers need not worry about making sure the connection is valid before invoking any vSphere API.

divyenpatel replied:
When VC client is created, we do not create CNS Client and PBM Client. They are getting created on-demand when required. 

`GetStoragePolicyIDByName()` was called multiple places and everywhere we were calling `err := vc.ConnectPbm(ctx)`. so moved this inside `GetStoragePolicyIDByName`. so the caller does not need to explictly call ConnectPbm before calling `GetStoragePolicyIDByName` method.

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -164,26 +171,26 @@ func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Man
+	info, found := volumeMigration.volumePathToVolumeID.Load(volumeSpec.VolumePath)
 	if found {
-		log.Infof("VolumeID: %q found from the cache for VolumePath: %q", info.(string), volumePath)
+		log.Infof("VolumeID: %q found from the cache for VolumePath: %q", info.(string), volumeSpec.VolumePath)
```
Info -> Debug

divyenpatel replied:
done.

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -164,26 +171,26 @@ func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Man
 		return "", err
 	}
-	log.Infof("successfully registered volume: %q with CNS. VolumeID: %q", volumePath, volumeID)
+	log.Infof("successfully registered volumeSpec: %v with CNS. VolumeID: %v", volumeSpec, volumeID)
```
successfully -> Successfully

divyenpatel replied:
done.

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -325,22 +332,22 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		return "", err
+	}
+	// Connect to vCenter
+	err = vCenter.Connect(ctx)
```
Why can this not be moved to VirtualCenter methods? We seem to be following this pattern where we rely on callers to make sure the connection is active/exists. I think we can reduce the complexity on the callers to do this by moving such things inside VirtualCenter methods.

divyenpatel replied:
moved `vCenter.Connect(ctx)` to VirtualCenter methods.

SandeepPissay commented on pkg/apis/migration/migration.go:
```diff
@@ -374,18 +392,24 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
 			},
 		}
-		log.Debugf("vSphere CNS driver registering volume %q with create spec %+v", volumePath, spew.Sdump(createSpec))
+		if storagePolicyID != "" {
```
We could move the createSpec variable creation outside the for loop and only change the BackingDiskUrlPath inside the loop. This will avoid garbage collection of many objects inside the loop.

divyenpatel replied:
yes moved createSpec out of the loop and only set BackingObjectDetails in the loop.

```
createSpec.BackingObjectDetails = &cnstypes.CnsBlockBackingDetails{BackingDiskUrlPath: backingDiskURLPath}
```

divyenpatel commented:
@SandeepPissay addressed your review comments as part of https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/287/commits/e9c7ae5add905d61ab6f0010778eb02611d5fb4e

divyenpatel commented:
/test pull-vsphere-csi-driver-unit-test

SandeepPissay commented:
@divyenpatel The code changes looks good. Can you do the following:
1. Run vanilla E2E tests to make sure that this code is not introducing any regression.
2. Sqash commit to a single commit and log message.

divyenpatel commented:
@SandeepPissay squashed all commits.
exected all block e2e tests.
Test log - https://gist.github.com/divyenpatel/bc07b5d2e31a9185c9a444e1302a528b

1 test failed out of 33 and failure is not related to this change.

```
Summarizing 1 Failure:

[Fail] [csi-block-vanilla] Relocate detached volume  [BeforeEach] Verify relocating detached volume works fine 
/Users/divyenp/go/src/github.com/divyenpatel/vsphere-csi-driver/tests/e2e/e2e_common.go:121

Ran 33 of 106 Specs in 6217.758 seconds
FAIL! -- 32 Passed | 1 Failed | 0 Pending | 73 Skipped
--- FAIL: TestE2E (6217.77s)
FAIL
```

SandeepPissay commented:
> @SandeepPissay squashed all commits.
> exected all block e2e tests.
> Test log - https://gist.github.com/divyenpatel/bc07b5d2e31a9185c9a444e1302a528b
> 
> 1 test failed out of 33 and failure is not related to this change.
> 
> ```
> Summarizing 1 Failure:
> 
> [Fail] [csi-block-vanilla] Relocate detached volume  [BeforeEach] Verify relocating detached volume works fine 
> /Users/divyenp/go/src/github.com/divyenpatel/vsphere-csi-driver/tests/e2e/e2e_common.go:121
> 
> Ran 33 of 106 Specs in 6217.758 seconds
> FAIL! -- 32 Passed | 1 Failed | 0 Pending | 73 Skipped
> --- FAIL: TestE2E (6217.77s)
> FAIL
> ```

Can you post this in the PR description?

divyenpatel commented:
> Can you post this in the PR description?

Done.

SandeepPissay commented:
/lgtm
